src/coreclr/gc/gc.cpp

```diff
@@ -20366,6 +20402,41 @@ void gc_heap::check_and_set_no_gc_oom()
 #endif //MULTIPLE_HEAPS
 }
 
+#ifdef USE_REGIONS
+int gc_heap::extend_soh_for_no_gc()
+{
+    size_t required = soh_allocation_no_gc;
+    heap_segment* region = generation_allocation_segment (generation_of (0));
+    uint8_t* allocated = (region == ephemeral_heap_segment) ? alloc_allocated : heap_segment_allocated (region);
+    size_t available = heap_segment_reserved (region) - allocated;
+    size_t commit = min (available, required);
+    if (grow_heap_segment (region, allocated + commit))
```

Comment:
there can be more existing regions in gen0 after ephemeral_heap_segment, those regions should be counted.